fix(core): disallow `(x**z*y**z) -> (x*y)**z` in as_base_exp

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

Without assumptions on x, y and z it does not hold that `(x**z)*(y**z)`
is equal to `(x*y)**z` so as_base_exp should not return `(x*y, z)`.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
   * `Mul.as_base_exp` no longer allows `(x**z*y**z) -> (x*y)**z` unless `z` is an integer. Previously this could lead to incorrect results if the bases are not positive and the exponent is not an integer.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
fix(core): disallow (x**z*y**z) -> (x*y)**z in as_base_exp

With assumptions on x, y and z it does not hold that `(x**z)*(y**z)`
is equal to `(x*y)**z` so as_base_exp should not return `(x*y, z)`.

oscarbenjamin committed:
fix(core): remove redundant as_base_exp methods

smichr commented on sympy/core/mul.py:
```diff
@@ -1245,7 +1245,7 @@ def as_base_exp(self):
             if e1 is None:
                 e1 = e
-            elif e != e1 or nc > 1:
+            elif e != e1 or nc > 1 or not e.is_Integer:
```
should that be `e.is_integer` to allow for `Symbol('e', integer=True)`?

oscarbenjamin replied:
No, we should not invoke the assumptions system in such a basic manipulation. I would rather remove this whole method but the change here is intended to be a minimal change that fixes the correctness problem.

oscarbenjamin committed:
integrals: Update Laplace transform tests

Previously Mul.as_base_exp allowed the following:

  >>> (sqrt(x)*sqrt(y)).as_base_exp()
  (x*y, S.Half)

This implicitly assumes that sqrt(x)*sqrt(y) == sqrt(x*y) which is not
true for all possible values of x and y e.g. if x = y = -1 then we have

  sqrt(-1)*sqrt(-1) = I*I = -1

However

  sqrt((-1)*(-1)) = sqrt(1) = 1

This change in as_base_exp means that lots of Laplace transform results
come out differently. This commit updates the tests to reflect this. I
have not verified that these changes are correct but the change to
as_base_exp should prevent as_base_exp itself from returning incorrect
results and the changes here are the effect that it has had on the
output from laplace_transform().

oscarbenjamin commented:
CC @hanspi42 

The change that I have made here fixes potential bugs from `as_base_exp` and has resulted in changing many of the outputs from `laplace_transform` for examples that are in the test suite. I have updated the tests to reflect this changed output but I have not verified that the changes are correct.

The differences are mostly (but not all) things like `sqrt(1/s)` vs `1/sqrt(s)`. These are not equivalent for negative or more generally for complex `s`. The bug in `as_base_exp` that is fixed here is that `as_base_exp` considered them equivalent. Consider:
```python
In [1]: x, y = symbols('x, y')

In [2]: expr = sqrt(x)*sqrt(y) - sqrt(x*y)

In [3]: expr
Out[3]: 
          _____
√x⋅√y - ╲╱ x⋅y 

In [4]: simplify(expr)
Out[4]: 
          _____
√x⋅√y - ╲╱ x⋅y 
```
It is correct that this does not simplify to zero because if `x` and `y` are negative for example then:
```python
In [5]: expr.subs({x:-1,y:-1})
Out[5]: -2
```
The bug in `as_base_exp` is this:
```python
In [6]: (sqrt(x)*sqrt(y)).as_base_exp()
Out[6]: (x⋅y, 1/2)

In [7]: (sqrt(x*y)).as_base_exp()
Out[7]: (x⋅y, 1/2)
```
This implies that the two expressions are equivalent and so any code that uses `as_base_exp` might mix these two expressions up when they are not equivalent.

I have fixed it in this PR:
```python
In [1]: x, y = symbols('x, y')

In [2]: (sqrt(x)*sqrt(y)).as_base_exp()
Out[2]: (√x⋅√y, 1)

In [3]: (sqrt(x*y)).as_base_exp()
Out[3]: (x⋅y, 1/2)
```
This change though seems to have resulted in many of the `laplace_transform` test cases producing different outputs. In principle the change to `as_base_exp` only prevents incorrect results but we now need to investigate carefully whether the changed outputs of the higher level `laplace_transform` function are correct.

The changed outputs are not equivalent to the previous outputs (`sqrt(1/s)` is not equivalent to `1/sqrt(s)`) so it cannot be the case that both the old and new output is correct for all possible values of the symbols like `a` and `s` that are in the test suite.

oscarbenjamin commented on sympy/core/tests/test_function.py:
```diff
@@ -568,8 +568,7 @@ def test_function__eval_nseries():
     assert expint(Rational(3, 2), -x)._eval_nseries(x, 5, None) == \
-        2 - 2*sqrt(pi)*sqrt(-x) - 2*x + x**2 + x**3/3 + x**4/12 + 4*I*x**(S(3)/2)*sqrt(-x)/3 + \
-        2*I*x**(S(5)/2)*sqrt(-x)/5 + 2*I*x**(S(7)/2)*sqrt(-x)/21 + O(x**5)
+        2 - 2*x - x**2/3 - x**3/15 - x**4/84 - 2*I*sqrt(pi)*sqrt(x) + O(x**5)
```
@anutosh491 does this look correct to you?

oscarbenjamin commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -214,16 +224,18 @@ def test_laplace_transform():
     assert (LT(erf(sqrt(a/t)/2), t, s, simplify=True) ==
             (1/s - exp(-sqrt(a)*sqrt(s))/s, 0, True))
     assert (LT(erfc(sqrt(a*t)), t, s, simplify=True) ==
-            (-sqrt(a)/(s*sqrt(a + s)) + 1/s, -a, True))
+            (-sqrt(a)/(s*sqrt(a + s)) + 1/s, 0, True))
```
I am not sure what caused this change...

oscarbenjamin commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -142,24 +141,35 @@ def test_laplace_transform():
+            + sqrt(a)*(-sqrt(pi)*erfc(sqrt(a)/sqrt(s))
+            + sqrt(s)*exp(-a/s)/sqrt(a))/(4*s**(S(3)/2)))*exp(a/s)), 0, True))
     assert (LT(sqrt(t)*cosh(2*sqrt(a*t)), t, s) ==
-            (sqrt(pi)*(a + s/2)*exp(a/s)/s**(S(5)/2), 0, True))
+            ((((sqrt(pi)*(2 - erfc(sqrt(a)/sqrt(s)))
+            + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2
+            - (-sqrt(pi)*erfc(sqrt(a)/sqrt(s))
+            + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2)*exp(a/s)
+            - sqrt(pi))/(4*s**(S(3)/2)) + (((sqrt(pi)*(2 - erfc(sqrt(a)/sqrt(s)))
+            + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2
+            - (-sqrt(pi)*erfc(sqrt(a)/sqrt(s))
+            + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2)*exp(a/s)
+            + sqrt(pi))/(4*s**(S(3)/2)), 0, True))
```
In some cases the changed output here is simpler but this case shows an example that is more complex. We need to verify correctness first and then think about improving simplicity afterwards.

oscarbenjamin replied:
In this example the output can simplify almost down to the previous result:
```python
In [10]: expr = (((sqrt(pi)*(2 - erfc(sqrt(a)/sqrt(s)))
    ...:             + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2
    ...:             - (-sqrt(pi)*erfc(sqrt(a)/sqrt(s))
    ...:             + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2)*exp(a/s)
    ...:             - sqrt(pi))/(4*s**(S(3)/2)) + (((sqrt(pi)*(2 - erfc(sqrt(a)/sqrt(s)))
    ...:             + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2
    ...:             - (-sqrt(pi)*erfc(sqrt(a)/sqrt(s))
    ...:             + sqrt(s)*exp(-a/s)/sqrt(a))*(2*a/s + 1)/2)*exp(a/s)
    ...:             + sqrt(pi))/(4*s**(S(3)/2))

In [11]: radsimp(expr).factor()
Out[11]: 
                      a
                ____  ─
               ╱  5   s
√π⋅(2⋅a + s)⋅╲╱  s  ⋅ℯ 
───────────────────────
            5          
         2⋅s  
```
The previous result was
```python
In [12]: sqrt(pi)*(a + s/2)*exp(a/s)/s**(S(5)/2)
Out[12]: 
            a
            ─
   ⎛    s⎞  s
√π⋅⎜a + ─⎟⋅ℯ 
   ⎝    2⎠   
─────────────
     5/2     
    s   
```
These are equivalent if `s^(5/2)` is equivalent to `s^5/sqrt(s^5)` but that doesn't hold. The SymPy interpretation of `s^(5/2)` is `(s^(1/5))^2` rather than `(s^5)^(1/2)` so these are not equivalent.

oscarbenjamin commented:
No need to rush. This bug has been around a long time. I'm surprised that fixing it hasn't caused more breakage.

By the way, I can't remember where we are at with the Laplace transform stuff. Are all the PRs merged or is there anything open and waiting?

Also as a secondary note I plan to fix the matrix exponential code soon which means making an improved implementation of inverse Laplace transform (with no Heavisides!) for rational functions. I will be interested to hear your opinions when it is ready.

hanspi42 commented:
> The differences are mostly (but not all) things like `sqrt(1/s)` vs `1/sqrt(s)`. These are not equivalent for negative or more generally for complex `s`.

This is a bit mind-boggling. I think it is like this: if we assume that we are allowed to write complex numbers as $s=Ae^{j\phi}$ and $-\frac{\pi}{2}<\phi<\frac{\pi}{2}$, and then define the main value of the square root as $s=\sqrt{A}e^{j\frac{\phi}{2}}$, then $\frac{1}{\sqrt{s}}=\sqrt{\frac{1}{s}}$ except if $s$ is negative and real. That last part is not a problem because if the Laplace transform contains something like this, the inverse Laplace transform will be a path integral whose path does go through the negative real axis. So in the context of the Laplace transform, $\frac{1}{\sqrt{s}}$ and $\sqrt{\frac{1}{s}}$ are interchangeable. Do you think this is correct?

oscarbenjamin commented:
> Do you think this is correct?

Maybe. In complex analysis it often makes more sense to think of e.g. `sqrt(s)` as a multivalued function rather than a branched function. SymPy doesn't treat it that way but it might be a correct interpretation for Laplace transforms.

In the specific case of `sqrt(1/s) == 1/sqrt(s)` it only fails for the negative reals right on the branch cut but other cases like `s == sqrt(s^2)` can fail for half or more of the complex plane. I'm not sure if there are examples like that here.

Note that we also have other kinds of sqrt like `sqrt(a)` where `a` is a different symbol (not the Laplace variable). Many of the cases here involve `sqrt(a)` appearing differently in the expression. Probably in context `a` is likely to be a real variable and we can't just dismiss the negative `a` case.

oscarbenjamin commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -214,16 +224,18 @@ def test_laplace_transform():
     # Error functions (laplace7.pdf)
     assert LT(erf(a*t), t, s) == (exp(s**2/(4*a**2))*erfc(s/(2*a))/s, 0, True)
-    assert LT(erf(sqrt(a*t)), t, s) == (sqrt(a)/(s*sqrt(a + s)), 0, True)
+    assert LT(erf(sqrt(a*t)), t, s) == (1/(s*sqrt(1 + s/a)), 0, True)
```
In this case we have:
```python
In [25]: a, s = symbols('a, s')

In [26]: e1 = sqrt(a)/(s*sqrt(a + s)) # old

In [27]: e2 = 1/(s*sqrt(1 + s/a)) # new

In [28]: e1
Out[28]: 
    √a     
───────────
    _______
s⋅╲╱ a + s 

In [29]: e2
Out[29]: 
      1      
─────────────
      _______
     ╱     s 
s⋅  ╱  1 + ─ 
  ╲╱       a 

In [30]: e1.subs(a, -1)
Out[30]: 
     ⅈ     
───────────
    _______
s⋅╲╱ s - 1 

In [31]: e2.subs(a, -1)
Out[31]: 
     1     
───────────
    _______
s⋅╲╱ 1 - s 
```
These are not equal for `s > 1`:
```python
In [32]: e1.subs(a, -1).subs(s, 2)
Out[32]: 
ⅈ
─
2

In [33]: e2.subs(a, -1).subs(s, 2)
Out[33]: 
-ⅈ 
───
 2 
```

hanspi42 replied:
Hmm, unfortunately Bateman54 gives the first formula, unless I made a stupid mistake. Is there a difference between $\sqrt{a}$ and $a^{1/2}$?
I hope I'm not wrong that $a^{1/2}t^{1/2}=(at)^{1/2}$ if t is positive and real...

oscarbenjamin replied:
`sqrt(a)` and `a^(1/2)` are the same thing.

It is also true that if `t` is positive and real then `sqrt(a*t) = sqrt(a)*sqrt(t)`

For `a = -1` Wolfram alpha gives a Piecewise for this case:
https://www.wolframalpha.com/input?i=LT%28erf%28sqrt%28-t%29%29%2C+t%2C+s%29

It is basically `Piecewise((e2, s<1), (e1, s>1)`. I think that if `a = -1` then the two formulae are equivalent for `s < 1` though so it should be possible to have `e1` at least for real `s`.

Is `e1` what you mean by the first formula? If so and if that is the formula in Bateman then how have we eneded with `e2`? Maybe a `sqrt` is being computed in the wrong place?

The formula here looks like `e1`:
https://github.com/sympy/sympy/blob/b354658bfd7b863ee59897321d9645efbb9d1f57/sympy/integrals/laplace.py#L487-L488
Is there something higher up that extracts the `a`?

hanspi42 replied:
Ah, I think I have it. In the test cases, `a` is defined as positive. With your new code,

```
a = symbols('a', real=True)
alpha = symbols('alpha', positive=True)
t = symbols('t', real=True)
print(sqrt(a*t))
print(sqrt(alpha*t))
```

immediately gives

```
sqrt(a*t)
sqrt(alpha)*sqrt(t)
```

before the LT can do anything, so my simple rule does not match anymore and the result is in the end achieved by integration. Something similar happens in all of the cases you cite.

For these test cases, the restriction to positive `a` in the tests was convenient such that not too many variables were needed, but it was not necessary. Now it bites us.

Howerver, we don't have a mathematical problem, since what happens here is correct for strictly positive `a`.

So I suggest the following:

Add
```
    ar, br, cr = symbols('ar, br, cr', real=True)
```
to the tests, and in all cases where an input like `sqrt(a*t)` fails, replace it by `sqrt(ar*t)`, and then it must give the same result as before your change to `as_base_exp`.

After you merge this, I will debug my pattern matching code to check why my simple rules cannot match the expression `erf(sqrt(alpha)*sqrt(t))` and instead defer them to integration.

oscarbenjamin replied:
Okay, so is it the case that with or without the change here in this PR the Laplace code does not match and the rule is not used?

Does that mean that actually the changed output results from the integration code being affected by the change to `as_base_exp()` rather than it affecting anything in the Laplace-specific code?

oscarbenjamin replied:
It would be good to use `ar` in the tests if `ar` has the real assumption set.

oscarbenjamin replied:
And `ap` if the positive assumption is set.

hanspi42 replied:
> Okay, so is it the case that with or without the change here in this PR the Laplace code does not match and the rule is not used?
> 
> Does that mean that actually the changed output results from the integration code being affected by the change to `as_base_exp()` rather than it affecting anything in the Laplace-specific code?

What happens with your change is that for positive `a` and in this PR, the rule is not used. It is used for real `a`. In sympy main it is used for both cases.

What happens in this PR is that for positive `a`, if you write `f=sqrt(a*t)`, the variable `f` will then contain `sqrt(a)*sqrt(t)`. Before this PR, the `sqrt` function did not do this.

hanspi42 replied:
> It would be good to use `ar` in the tests if `ar` has the real assumption set.

True, so you could now, for this PR, implement the `ar`, and I will make a second update later to make the test code more readable by using `ap` only where it is really needed.

oscarbenjamin commented:
> What happens with your change is that for positive a and in this PR, the rule is not used. It is used for real a. In sympy main it is used for both cases.

Ah, okay. I think I see why this is showing up for the Laplace tests then. It is because Laplace makes heavy use of pattern-matching and `Pow.matches` uses `as_base_exp`:
https://github.com/sympy/sympy/blob/b354658bfd7b863ee59897321d9645efbb9d1f57/sympy/core/power.py#L1440-L1454
The change here causes some patterns to not match in the Laplace code which means that the transforms are computed by the integration routines instead of the rule-based transform.

The differences seen here are all differences between the integration code and the rule-based transform. Those differences are still worth investigating to see if they are bugs in the integration code or the Laplace rules.

The Laplace rules probably need to be augmented then so that a rule like:
```
erf(a*sqrt(t)) -> f(a)
```
should be changed to:
```
erf(a*sqrt(b*t)) -> f(a,b)
```

oscarbenjamin commented:
I'm rerunning the tests.

Now that we understand what has caused the change in the Laplace code I think that this is good to go. Future improvements will be needed there but we should get this merged first.

This is ready for review.

hanspi42 commented:
I think you misunderstood what I meant. You should re-write the `test_laplace_transform()` such that no big changes happen in this PR. I could do this for you, should I just write it and copy-paste here, or what would be the proper way to do it with git?

hanspi42 commented on sympy/integrals/tests/test_laplace.py:
```diff
@@ -142,24 +141,35 @@ def test_laplace_transform():
+            (sqrt(pi)*exp(a/s)*erf(sqrt(a)/sqrt(s))/sqrt(s), 0, True))
     assert (LT(sinh(sqrt(a*t))**2/sqrt(t), t, s) ==
-            (sqrt(pi)*(exp(a/s) - 1)/(2*sqrt(s)), 0, True))
+            (sqrt(pi)*(2*exp(a/s) - 2)/(4*sqrt(s)), 0, True))
```
Please don't change the test cases like this. This should be, for example,

```
assert (LT(sinh(sqrt(ar*t))**2/sqrt(t), t, s) ==
    (sqrt(pi)*(exp(a/s) - 1)/(2*sqrt(s)), Max(0, -a), True))
```

or similar.

oscarbenjamin commented:
> You should re-write the `test_laplace_transform()` such that no big changes happen in this PR

Okay, I see what you mean.

In that case only the first two commits from this PR are needed.

> should I just write it and copy-paste here, or what would be the proper way to do it with git?

Probably the best thing is just to open a new PR starting with the first two commits here and then adding a commit that fixes the Laplace tests.

The way that I do this is like this:

Add something to your git config so that git fetch brings down pull requests:
https://gist.github.com/piscisaureus/3342247

Mine is:
```
[remote "upstream"]
	url = git@github.com:sympy/sympy.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
	fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*
	gh-resolved = base
```
(I'm actually not sure what the `gh-resolved` line does.)

Now you can do `git fetch upstream` which will download all pull requests.

Now you can checkout this pull request (minus the last commit: `^`):
```
git checkout upstream/pr/26885^
```
Now make a new branch:
```
git checkout -b my_branch
```
At this point you can fix up the Laplace tests, commit and open a new PR.

hanspi42 commented:
> The way that I do this is like this:

OK, no, that is too much effort for a relatively small change. I'll just send you the file such that you can copy it into place, and then I'll restructure the tests in a new PR of my own after you have merged this.

I'll probably be able to do it on Friday.

oscarbenjamin commented:
A simpler approach is just to copy the non-Laplace changes out if you would rather not do it with git.

All I have done here is delete two small functions and change 4 lines of code.

oscarbenjamin commented:
The reason I suggested this approach:

> The way that I do this is like this:
> 
> Add something to your git config

Is just because it makes it generally easy to checkout a pull request e.g. I can always check out pull request number 123 just like:
```
git checkout upstream/pr/123
```
A more direct way if you don't want to mess with git config is just this:
```
git fetch upstream pull/26885/head:oscar_branch
git checkout oscar_branch^
git checkout -b my_branch
```
Then you have the two commits from this PR and you can run the unchanged Laplace tests (which will fail without the third commit). You can then make changes and open a PR from my_branch.

hanspi42 commented:
Hi @oscarbenjamin, I will have to look into what Bateman did in 1954, for the time being I have commented out all rules with `sqrt(a*t)` and `sqrt(a/t)`. I consider that Bateman's tables could be wrong in these cases. I have also commented out all test cases because they test whether said rules trigger and are not mathematically necessary. See new PR #26965
